### PR TITLE
graalvm-{8,11}-ce: reduce closure sizes on Linux for packages that builds on it

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

Nowadays most packages builds from GraalVM Native Image have huge closures. For example, `babashka`:

```
$ nix path-info -S ./result -h
/nix/store/yfnnfs75vh01sw6aqcrix1rgk919c4sb-babashka-0.6.1	  2.0G
```

The reason for this is that any GraalVM Native Image package links to C libraries that are located inside GraalVM closure:

```
$ ldd ./result/bin/bb
	linux-vdso.so.1 (0x00007ffe9bf56000)
	libstdc++.so.6 => /nix/store/bg7pq2k5i1ylpq5czsh7zym5nijpj30q-gcc-10.3.0-lib/lib/libstdc++.so.6 (0x00007f8e7a926000)
	libm.so.6 => /nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce/lib/svm/clibraries/linux-amd64/libm.so.6 (0x00007f8e7a7e5000)
	libpthread.so.0 => /nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce/lib/svm/clibraries/linux-amd64/libpthread.so.0 (0x00007f8e7a7c5000)
	libdl.so.2 => /nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce/lib/svm/clibraries/linux-amd64/libdl.so.2 (0x00007f8e7a7c0000)
	librt.so.1 => /nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce/lib/svm/clibraries/linux-amd64/librt.so.1 (0x00007f8e7a7b5000)
	libc.so.6 => /nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce/lib/svm/clibraries/linux-amd64/libc.so.6 (0x00007f8e7a5ee000)
	/nix/store/2jfn3d7vyj7h0h6lmh510rz31db68l1i-glibc-2.33-50/lib/ld-linux-x86-64.so.2 => /nix/store/jsp3h3wpzc842j0rz61m5ly71ak6qgdn-glibc-2.32-54/lib64/ld-linux-x86-64.so.2 (0x00007f8e7aafd000)
	libgcc_s.so.1 => /nix/store/bg7pq2k5i1ylpq5czsh7zym5nijpj30q-gcc-10.3.0-lib/lib/libgcc_s.so.1 (0x00007f8e7a5d4000)

$ nix path-info -S /nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce -h
/nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce	  1.9G
```

However, if you look at the `graalvm-*-ce` derivations, you can see that those C libraries are just `glibc` (and some `glibc.static` and `zlib.static`, but they only matter for static compilation):

https://github.com/NixOS/nixpkgs/blob/ac27d4a7a48fd5223fb18c205fa2cfbe44189f4d/pkgs/development/compilers/graalvm/community-edition.nix#L173-L176

To fix it, we could add `-H:CLibraryPath=${glibc}/lib` to make GraalVM to depend on the libraries from `glibc` instead their internal copies. However this approach is fragile. If for some reason GraalVM adds more libraries as dependencies, we will need to go back and fix all packages that depends on it.

So this PR introduces another approach: it adds a `lib` output to `graalvm-{8,11}-ce` and adds `-H:CLibraryPath=${graalvm-11-ce}/lib` to all(?) packages that builds from it. `lib` for now it is just `glibc` (the only dynamic dependency of GraalVM, the other ones are static ones). GraalVM itself always include the default C library location  (e.g.: `-H:CLibraryPath=${graalvm-11-ce}/lib/svm/clibraries/${platform}`) on its path, so this shouldn't break anything.

The above changes result in:

```
$ nix path-info -S results/babashka -h
/nix/store/fay0p3hiiic4kd4fk0xyczl1ff8sjcp1-babashka-0.6.1	116.6M

$ ldd results/babashka/bin/bb
	linux-vdso.so.1 (0x00007ffe2737d000)
	libstdc++.so.6 => /nix/store/bg7pq2k5i1ylpq5czsh7zym5nijpj30q-gcc-10.3.0-lib/lib/libstdc++.so.6 (0x00007f7deef0a000)
	libm.so.6 => /nix/store/9qf2plvypbk93q2gw742dr0wdkw8jqj0-graalvm11-ce-lib/lib/libm.so.6 (0x00007f7deedc9000)
	libpthread.so.0 => /nix/store/9qf2plvypbk93q2gw742dr0wdkw8jqj0-graalvm11-ce-lib/lib/libpthread.so.0 (0x00007f7deeda9000)
	libdl.so.2 => /nix/store/9qf2plvypbk93q2gw742dr0wdkw8jqj0-graalvm11-ce-lib/lib/libdl.so.2 (0x00007f7deeda4000)
	librt.so.1 => /nix/store/9qf2plvypbk93q2gw742dr0wdkw8jqj0-graalvm11-ce-lib/lib/librt.so.1 (0x00007f7deed99000)
	libc.so.6 => /nix/store/9qf2plvypbk93q2gw742dr0wdkw8jqj0-graalvm11-ce-lib/lib/libc.so.6 (0x00007f7deebd2000)
	/nix/store/2jfn3d7vyj7h0h6lmh510rz31db68l1i-glibc-2.33-50/lib/ld-linux-x86-64.so.2 => /nix/store/2jfn3d7vyj7h0h6lmh510rz31db68l1i-glibc-2.33-50/lib64/ld-linux-x86-64.so.2 (0x00007f7def0e1000)
	libgcc_s.so.1 => /nix/store/bg7pq2k5i1ylpq5czsh7zym5nijpj30q-gcc-10.3.0-lib/lib/libgcc_s.so.1 (0x00007f7deebb8000)

$ nix path-info -S /nix/store/9qf2plvypbk93q2gw742dr0wdkw8jqj0-graalvm11-ce-lib -h
/nix/store/9qf2plvypbk93q2gw742dr0wdkw8jqj0-graalvm11-ce-lib	 31.6M
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by thiagokokada — COMMENTED** [review]
You can see the closure reductions on each commit. But they're all very significant (from around 2GB to 70~120MB).

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 140287` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>graalvm11-ce</li>
    <li>graalvm8-ce</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Sorry for the huge amount of review requests, it is just that this PR changes many packages.

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 140287` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>graalvm11-ce</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

File: pkgs/development/tools/misc/clojure-lsp/default.nix
Comment (ericdallo) [inline pkgs/development/tools/misc/clojure-lsp/default.nix:32]: clojure-lsp already has this flag built-in on its graal config, so I don't think we need these lines, right?
https://github.com/clojure-lsp/clojure-lsp/blob/master/resources/META-INF/native-image/clojure-lsp/clojure-lsp/native-image.properties#L6

**Review by ericdallo — APPROVED** [review]
LGTM!

File: pkgs/development/tools/misc/clojure-lsp/default.nix
Comment (thiagokokada) [reply pkgs/development/tools/misc/clojure-lsp/default.nix]: Removed at commit d672332 and still working:

Result of `nixpkgs-review pr 140287` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>graalvm11-ce</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Considering that @ericdallo is the maintainer for graalvm and clojure-lsp, I just need a review from either @jlesquembre or @bennyandresen (for babashka and clj-kondo) and @stelcodes for zprint.

**Comment by thiagokokada** [comment]
Thanks @SuperSandro2000, you rock!

**Pull request merged**
